src/test/TestUtils.cpp

```diff
@@ -194,6 +197,78 @@ genesis(int minute, int second)
         getTestDateTime(1, 7, 2014, 0, minute, second));
 }
 
+ConfigUpgradeSetKey
+upgradeSorobanNetworkConfig(std::function<void(SorobanNetworkConfig&)> modifyFn,
+                            SimulationPtr simulation)
+{
+    auto nodes = simulation->getNodes();
+    auto& lg = nodes[0]->getLoadGenerator();
+    auto& app = *nodes[0];
+
+    auto& complete =
+        app.getMetrics().NewMeter({"loadgen", "run", "complete"}, "run");
+    auto completeCount = complete.count();
+    // Only create an account if there are none aleady created.
+    if (app.getMetrics()
+            .NewMeter({"loadgen", "account", "created"}, "account")
+            .count() == 0)
+    {
+        auto createAccountsLoadConfig =
+            GeneratedLoadConfig::createAccountsLoad(1, 1);
+        createAccountsLoadConfig.offset =
+            std::numeric_limits<uint32_t>::max() - 1;
+        lg.generateLoad(createAccountsLoadConfig);
+        simulation->crankUntil(
+            [&]() { return complete.count() == completeCount + 1; },
+            300 * Herder::EXP_LEDGER_TIMESPAN_SECONDS, false);
+    }
+
+    // Create upload wasm transaction
+    auto createUploadCfg = GeneratedLoadConfig::createSorobanUpgradeSetupLoad();
+    createUploadCfg.offset = std::numeric_limits<uint32_t>::max() - 1;
+    lg.generateLoad(createUploadCfg);
+    completeCount = complete.count();
+    simulation->crankUntil(
+        [&]() { return complete.count() == completeCount + 1; },
+        300 * Herder::EXP_LEDGER_TIMESPAN_SECONDS, false);
+
+    // Create upgrade transaction.
+    auto createUpgradeLoadGenConfig = GeneratedLoadConfig::txLoad(
+        LoadGenMode::SOROBAN_CREATE_UPGRADE, 1, 1, 1);
+    createUpgradeLoadGenConfig.offset =
+        std::numeric_limits<uint32_t>::max() - 1;
+    SorobanNetworkConfig cfg;
+    modifyFn(cfg);
+    createUpgradeLoadGenConfig.copySorobanNetworkConfigToUpgradeConfig(cfg);
+    auto upgradeSetKey = lg.getConfigUpgradeSetKey(createUpgradeLoadGenConfig);
+    lg.generateLoad(createUpgradeLoadGenConfig);
+    completeCount = complete.count();
+    simulation->crankUntil(
+        [&]() { return complete.count() == completeCount + 1; },
+        300 * Herder::EXP_LEDGER_TIMESPAN_SECONDS, false);
+
+    // Arm for upgrade.
+    for (auto app : nodes)
+    {
+        Upgrades::UpgradeParameters scheduledUpgrades;
+        auto lclHeader =
+            app->getLedgerManager().getLastClosedLedgerHeader().header;
+        scheduledUpgrades.mUpgradeTime =
+            VirtualClock::from_time_t(lclHeader.scpValue.closeTime);
+        scheduledUpgrades.mConfigUpgradeSetKey = upgradeSetKey;
+        app->getHerder().setUpgrades(scheduledUpgrades);
+    }
+    // Wait for upgrade to be applied
+    simulation->crankUntil(
+        [&]() {
+            auto status = app.getStatusManager().getStatusMessage(
```

Comment:
This doesn't properly wait for the upgrade to go through: setUpgrades sets status message REQUIRES_UPGRADES immediately, so this call to crankUntil is essentially a no-op. This is probably why some of the tests are failing. 

I'd suggest cranking until `cfg` equals config you get from `LedgerManager.getSorobanNetworkConfig`. Basically, you need to ensure that upgrades were applied to the ledger.

Reply:
As I understand it, `REQUIRES_UPGRADES` is a category of status message, `setUpgrades` [ sets the contents ](https://github.com/stellar/stellar-core/blob/master/src/herder/HerderImpl.cpp#L1513) of the `REQUIRES_UPGRADES` status message to "Armed with network upgrades...". When the upgrades are applied, the `REQUIRES_UPGRADES` status message is removed. This cranks until that status message is empty. I'm fairly certain this works as I have [assertions on the upgraded values after this completes](https://github.com/stellar/stellar-core/pull/4413/files/7275997f7876adcbc806bc99772347f0f30f8b71#diff-519110afbb88598be9bd6ca994ca41c4a80840b66febd9d58765ca96955c3389R122), but I can change it to compare with the config from `LedgerManager.getSorobanNetworkConfig`.

Reply:
I see. I believe relying on status message still isn't enough to tell that upgrades went through, for a few reasons:
* Core clears upgrades when it reaches consensus on a block containing upgrades, but _before_ the block is applied. This means that your crank function is potentially racing with the actual upgrade application, so we don't know if upgrades were applied when this function exits. 
* It seems a bit error-prone to rely on an empty status message: core is a state machine so it might enter a different state, and set status to something else, which would cause the simulation to time out.

I think the most reliable approach would be to rely directly on what's in the database, e.g. use `getSorobanNetworkConfig`